spring-data-relational/src/main/java/org/springframework/data/relational/core/mapping/AggregatePath.java

```diff
@@ -0,0 +1,467 @@
+
+		this.context = context;
+		this.path = path;
+
+		this.rootType = null;
+	}
+
+	public AggregatePath(RelationalMappingContext context, RelationalPersistentEntity<?> rootType) {
+
+		Assert.notNull(context, "context must not be null");
+		Assert.notNull(rootType, "rootType must not be null");
+
+		this.context = context;
+		this.rootType = rootType;
+
+		this.path = null;
+	}
+
+	public static boolean isWritable(@Nullable PersistentPropertyPath<? extends RelationalPersistentProperty> path) {
+		return path == null || path.getLeafProperty().isWritable() && isWritable(path.getParentPath());
+	}
+
+	public boolean isRoot() {
+		return path == null;
+	}
+
+	/**
+	 * Tests if {@code this} and the argument represent the same path.
+	 *
+	 * @param path to which this path gets compared. May be {@literal null}.
+	 * @return Whence the argument matches the path represented by this instance.
+	 */
+	public boolean matches(@Nullable PersistentPropertyPath<RelationalPersistentProperty> path) {
+		return this.path == null ? path == null || path.isEmpty() : this.path.equals(path);
+	}
+
+	/**
+	 * The name of the column used to reference the id in the parent table.
+	 *
+	 * @throws IllegalStateException when called on an empty path.
+	 */
+	public SqlIdentifier getReverseColumnName() {
+
+		Assert.state(path != null, "Empty paths don't have a reverse column name");
+
+		return path.getLeafProperty().getReverseColumnName(this);
+	}
+
+	/**
+	 * Returns the path that has the same beginning but is one segment shorter than this path.
+	 *
+	 * @return the parent path. Guaranteed to be not {@literal null}.
+	 * @throws IllegalStateException when called on an empty path.
+	 */
+	public AggregatePath getParentPath() {
+
+		if (path == null) {
+			throw new IllegalStateException("The parent path of a root path is not defined.");
+		}
+
+		if (path.getLength() == 1) {
+			return context.getAggregatePath(path.getLeafProperty().getOwner());
+		}
+
+		return context.getAggregatePath(path.getParentPath());
+	}
+
+	/**
+	 * The {@link RelationalPersistentEntity} associated with the leaf of this path.
+	 *
+	 * @return Might return {@literal null} when called on a path that does not represent an entity.
+	 */
+	@Nullable
+	public RelationalPersistentEntity<?> getLeafEntity() {
+		return path == null ? rootType : context.getPersistentEntity(path.getLeafProperty().getActualType());
+	}
+
+	/**
+	 * @return {@literal true} if this path represents an entity which has an Id attribute.
+	 */
+	public boolean hasIdProperty() {
+
+		RelationalPersistentEntity<?> leafEntity = getLeafEntity();
+		return leafEntity != null && leafEntity.hasIdProperty();
+	}
+
+	/**
+	 * Returns the longest ancestor path that has an {@link org.springframework.data.annotation.Id} property.
+	 *
+	 * @return A path that starts just as this path but is shorter. Guaranteed to be not {@literal null}.
+	 */
+	public AggregatePath getIdDefiningParentPath() {
```

Comment:
It would be neat to refactor this method to a path-traversal method to walk the path hierarchy with a `Predicate`. The concrete functionality here would fit into a utility as paths should not be tied to identifiers.